### PR TITLE
sgrep_lint: Pass through include/exclude flags to sgrep

### PR DESCRIPTION
**Pull request by underyx**

### CODE REVIEW

File: sgrep_lint/sgrep_main.py [hunk 1]
```diff
@@ -115,11 +114,25 @@ def sgrep_error_json_to_message_then_exit(
...
+    """Yields include/exclude CLI options to call semgrep with.
+
+    This is based on the arguments given to semgrep-lint.
+    """
+    for pattern in args.include:
+        yield from ["-include", pattern]
+    for pattern in args.exclude:
+        yield from ["-exclude", pattern]
+    for pattern in args.exclude_dir:
+        yield from ["-exclude-dir", pattern]
```
Comment (underyx) [inline sgrep_lint/sgrep_main.py:127]: Couldn't make up my mind whether this is elegant, or gratuitously magical, instead of just appending to a list.

**Comment by DrewDennison** [comment]
@brendongo thoughts on if this passes your quality bar for develop or would prefer to see in a coinbase branch?

File: sgrep_lint/sgrep_main.py (hunk 1 above)
Comment (DrewDennison) [reply sgrep_lint/sgrep_main.py:127]: I think this is fine but model @brendongo as the gatekeeper of magic 🧙‍♂️

**Comment by brendongo** [comment]
Exclude should just work with directories no? i.e. why is there --exclude and --exclude-dirs

**Comment by underyx** [comment]
> Exclude should just work with directories no? i.e. why is there --exclude and --exclude-dirs

It's mirroring this from grep:

```
     --include
             If specified, only files matching the given filename pattern are searched.  Note that --exclude patterns take prior-
             ity over --include patterns.  Patterns are matched to the full path specified, not only to the filename component.

     --include-dir
             If -R is specified, only directories matching the given filename pattern are searched.  Note that --exclude-dir pat-
             terns take priority over --include-dir patterns.

     --exclude
             If specified, it excludes files matching the given filename pattern from the search.  Note that --exclude patterns
             take priority over --include patterns, and if no --include pattern is specified, all files are searched that are not
             excluded.  Patterns are matched to the full path specified, not only to the filename component.

     --exclude-dir
             If -R is specified, it excludes directories matching the given filename pattern from the search.  Note that
             --exclude-dir patterns take priority over --include-dir patterns, and if no --include-dir pattern is specified, all
             directories are searched that are not excluded.
```

**Comment by underyx** [comment]
> Exclude should just work with directories no? i.e. why is there --exclude and --exclude-dirs

Also https://github.com/returntocorp/enterprise/issues/55 requests this specifically in the issue description, so sgrep-core implemented these three flags.

**Comment by brendongo** [comment]
Do we have UX thinking around this. While I see the value in mimicing grep options not sure we want to do so without proper thinking. IMO there should be a way to exclude a file and there should only be one way to exclude a file. This seems like it will just be confusing. @dlukeomalley

**Comment by DrewDennison** [comment]
yeah this commands names look good to me and I'm sure @dlukeomalley did some product thinking there in creating the ticket based on grep

**Review by DrewDennison — DISMISSED** [review]

**Comment by underyx** [comment]
FWIW design discussion is on https://github.com/returntocorp/enterprise/issues/55

**Comment by brendongo** [comment]
This is also missing --include-dirs

**Comment by brendongo** [comment]
> > Exclude should just work with directories no? i.e. why is there --exclude and --exclude-dirs
> 
> Also [returntocorp/enterprise#55](https://github.com/returntocorp/enterprise/issues/55) requests this specifically in the issue description, so sgrep-core implemented these three flags.

I don't see any discussion on there regarding having a single exclude vs having an exclude and exclude-dir?

**Review by brendongo — CHANGES_REQUESTED** [review]
waiting on product and/or --include-dir

File: sgrep_lint/sgrep.py [hunk 2]
```diff
@@ -62,11 +62,24 @@
         help=f"only invoke sgrep if config(s) are valid",
         action="store_true",
     )
+
+    parser.add_argument(
+        "--include",
+        action="append",
+        default=[],
+        help="Scan only files with this filename. Can be added multiple times to include multiple patterns.",
```
Comment (ievans) [inline sgrep_lint/sgrep.py]: is this a filename, or a path pattern? An example might be nice in the helptext too

File: sgrep_lint/sgrep.py [hunk 3]
```diff
@@ -62,11 +62,24 @@
...
+    )
     parser.add_argument(
         "--exclude",
         action="append",
         default=[],
-        help="Path pattern to exclude. Can be added multiple times to exclude multiple patterns.",
+        help="Don't scan files with this filename. Can be added multiple times to exclude multiple patterns.",
```
Comment (ievans) [inline sgrep_lint/sgrep.py]: same comment: is this a filename, or a path pattern? An example might be nice in the helptext too

**Comment by underyx** [comment]
> waiting on product and/or --include-dir

Note that `--include-dir` is not yet supported by sgrep-core, so if we decide on that, it will go to @aryx first.

File: sgrep_lint/tests/run-lint-tests.sh [hunk 4]
```diff
@@ -109,6 +142,9 @@ local_tests() {
     test_registry
     test_sgrep_default_file
     test_sgrep_default_folder
+    test_sgrep_include
+    test_sgrep_exclude
+    test_sgrep_exclude_dir
```
Comment (ievans) [inline sgrep_lint/tests/run-lint-tests.sh:147]: v nice!

File: sgrep_lint/sgrep.py (hunk 2 above)
Comment (underyx) [reply sgrep_lint/sgrep.py]: Filename, noted, thanks, adding an example!

File: sgrep_lint/sgrep.py [hunk 5]
```diff
@@ -62,11 +62,24 @@
         help=f"only invoke sgrep if config(s) are valid",
         action="store_true",
     )
+
+    parser.add_argument(
+        "--include",
+        action="append",
+        default=[],
+        help="Scan only files with this filename, such as --include='*.js' Add multiple times to include multiple patterns.",
```
Comment (underyx) [inline sgrep_lint/sgrep.py:70]: @ievans can you check these examples? ☺️

File: sgrep_lint/tests/run-lint-tests.sh (hunk 4 above)
Comment (underyx) [reply sgrep_lint/tests/run-lint-tests.sh:147]: yay 🙃

**Comment by underyx** [comment]
Note that due to our time zone difference, if merging is blocked on copy changes, typo fixes, or something small like that, feel free to edit my branch.

File: sgrep_lint/sgrep.py (hunk 5 above)
Comment (ievans) [reply sgrep_lint/sgrep.py:70]: It's an improvement! but could we a bit more explicit about what exactly is matched? Eg compare with man grep:
```
     --exclude
             If specified, it excludes files matching the given filename pattern from the search.  Note that --exclude patterns take priority over --include patterns, and if no --include pattern is specified, all files are searched that are not excluded.  Patterns are matched to the full path specified, not only to the filename component.
```
Is that last bit true here or no?

**Comment by DrewDennison** [comment]
updated the eng plan and moving @brendongo to be the decision maker here. I'll update the PR to remove --exclude-er

**Comment by brendongo** [comment]
Can we also add a test that --exclude and --include work on directories?

**Pull request closed without merging**